File: docs/examples/cplusplus/sensitivity_analysis.cpp
/*
 *    MIT No Attribution
 *
 *    Copyright 2023 Joel Andersson, Joris Gillis, Moritz Diehl, KU Leuven.
 *
 *    Permission is hereby granted, free of charge, to any person obtaining a copy of this
 *    software and associated documentation files (the "Software"), to deal in the Software
 *    without restriction, including without limitation the rights to use, copy, modify,
 *    merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
 *    permit persons to whom the Software is furnished to do so.
 *
 *    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 *    INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 *    PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 *    HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 *    OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 *    SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 */

#include <casadi/casadi.hpp>

#include <iostream>
#include <iomanip>

using namespace casadi;

struct Test {
  SXDict dae;
  double tf;
  std::vector<double> x0;
  double u0;
  bool is_ode;
  std::string name;
};

/** \brief Generate a simple ODE */
Test simpleODE(){
  // Time
  SX t = SX::sym("t");

  // Parameter
  SX u = SX::sym("u");

  // Differential states
  SX s = SX::sym("s"), v = SX::sym("v"), m = SX::sym("m");
  SX x = vertcat(s, v, m);

  // Constants
  double alpha = 0.05; // friction
  double beta = 0.1;   // fuel consumption rate

  // Differential equation
  SX ode = vertcat(v, (u-alpha*v*v)/m, -beta*u*u);

  // Quadrature
  SX quad = pow(v,3) + pow((3-sin(t))-u,2);

  // Return test problem
  Test r;
  r.dae = decltype(r.dae){{"t", t}, {"x", x}, {"p", u}, {"ode", ode}, {"quad", quad}}; // decltype because MSVC bug
  r.tf = 0.5;
  r.x0 = {0, 0, 1};
  r.u0 = 0.4;
  r.is_ode = true;
  r.name = "simple_ode";
  return r;
}

/** \brief Generate a simple DAE */
Test simpleDAE(){
  // Parameter
  SX u = SX::sym("u");

  // Differential state
  SX x = SX::sym("x");

  // Algebraic variable
  SX z = SX::sym("z");

  // Differential equation
  SX ode = -x + 0.5*x*x + u + 0.5*z;

  // Algebraic constraint
  SX alg = z + exp(z) - 1. + x;

  // Quadrature
  SX quad = x*x + 3.0*u*u;

  // Return DAE
  Test r;
  r.dae = decltype(r.dae){{"x", x}, {"z", z}, {"p", u}, {"ode", ode}, {"alg", alg}, {"quad", quad}};
  r.tf = 5;
  r.x0 = {1};
  r.u0 = 0.4;
  r.is_ode = false;
  r.name = "simple_dae";
  return r;
}

struct Solver {
  std::string plugin;
  bool ode_only;
  Dict opts;
};

int main(){

  // Test problems
  std::vector<Test> tests = {simpleODE(), simpleDAE()};

  // ODE/DAE integrators
  std::vector<Solver> solvers;
  solvers.push_back({"cvodes", true, Dict()});
  solvers.push_back({"idas", false, Dict()});
  solvers.push_back({"rk", true, Dict()});
  Dict kinsol_options = {{"linear_solver_type", "dense"}};

  Dict coll_opts = {{"rootfinder", "kinsol"},
                    {"collocation_scheme", "legendre"},
                    {"rootfinder_options", kinsol_options}};
  solvers.push_back({"collocation", false, coll_opts});

  // Loop over all problems
  for (auto&& test : tests) {
    // Loop over all solvers
    for (auto&& solver : solvers) {
      // Test with or without multiple output times
      for (int ntout : {1, 3}) {
        // Skip if problem cannot be handled
        if (solver.ode_only && !test.is_ode) continue;
        // Printout
        std::cout << "Solving \"" << test.name << "\" using \"" << solver.plugin << "\", " << ntout << " output times" << std::endl;
        // Output time grid
        std::vector<double> tout;
        if (ntout == 1) {
          tout = {test.tf};
        } else {
          tout = {0, 0.5 * test.tf, test.tf};
        }
        // Create integrator instance
        Function I = integrator("I", solver.plugin, test.dae, 0, tout, solver.opts);

        // Buffers for evaluation
        std::map<std::string, DM> arg, res;

        // Integrate to get results
        arg = decltype(arg){{"x0", test.x0},
                            {"p", test.u0}};
        res = I(arg);
        DM xf(res.at("xf"));
        DM qf(res.at("qf"));
        std::cout << std::setw(50) << "Unperturbed solution: " << "xf  = " << xf.nonzeros() <<  ", qf  = " << qf.nonzeros() << std::endl;

        // Perturb solution to get a finite difference approximation
        double h = 0.001;
        arg["p"] = test.u0+h;
        res = I(arg);
        DM fd_xf = (res.at("xf")-xf)/h;
        DM fd_qf = (res.at("qf")-qf)/h;
        std::cout << std::setw(50) << "Finite difference approximation: " << "d(xf)/d(p) = " << fd_xf.nonzeros() << ", d(qf)/d(p) = " << fd_qf.nonzeros() << std::endl;

        // Calculate once, forward
        Function I_fwd = I.factory("I_fwd", {"x0", "p", "fwd:x0", "fwd:p"},
                                            {"fwd:xf", "fwd:qf"});
        arg = decltype(arg){{"x0", test.x0},
                            {"p", test.u0},
                            {"fwd_x0", 0},
                            {"fwd_p", 1}};
        res = I_fwd(arg);
        DM fwd_xf = res.at("fwd_xf");
        DM fwd_qf = res.at("fwd_qf");
        std::cout << std::setw(50) << "Forward sensitivities: " << "d(xf)/d(p) = " << fwd_xf.nonzeros() << ", d(qf)/d(p) = " << fwd_qf.nonzeros() << std::endl;

        // Calculate once, adjoint
        std::vector<double> adj_qf(ntout, 0);
        adj_qf.back() = 1;
        Function I_adj = I.factory("I_adj", {"x0", "p", "adj:xf", "adj:qf"},
                                            {"adj:p", "adj:x0"});
        arg = decltype(arg){{"x0", test.x0}, {"p", test.u0}, {"adj_xf", 0}, {"adj_qf", adj_qf}};
        res = I_adj(arg);
        DM adj_x0 = res.at("adj_x0");
        DM adj_p = res.at("adj_p");
        std::cout << std::setw(50) << "Adjoint sensitivities: " << "d(qf)/d(x0) = " << adj_x0.nonzeros() << ", d(qf)/d(p) = " << adj_p.nonzeros() << std::endl;

        // Perturb adjoint solution to get a finite difference approximation of the second order sensitivities
        arg["p"] = test.u0+h;
        res = I_adj(arg);
        DM fd_adj_x0 = (res.at("adj_x0")-adj_x0)/h;
        DM fd_adj_p = (res.at("adj_p")-adj_p)/h;
        std::cout << std::setw(50) << "FD of adjoint sensitivities: " << "d2(qf)/d(x0)d(p) = " << fd_adj_x0.nonzeros() << ", d2(qf)/d(p)d(p) = " << fd_adj_p.nonzeros() << std::endl;

        // Forward over adjoint to get the second order sensitivities
        Function I_foa = I_adj.factory("I_foa",
                                      {"x0", "p", "fwd:p", "adj_xf", "adj_qf"},
                                      {"fwd:adj_x0", "fwd:adj_p"});
        arg = decltype(arg){{"x0", test.x0},
                            {"p", test.u0},
                            {"fwd_p", 1},
                            {"adj_xf", 0},
                            {"adj_qf", adj_qf}};
        res = I_foa(arg);
        DM fwd_adj_x0 = res.at("fwd_adj_x0");
        DM fwd_adj_p = res.at("fwd_adj_p");
        std::cout << std::setw(50) << "Forward over adjoint sensitivities: " << "d2(qf)/d(x0)d(p) = " << fwd_adj_x0.nonzeros() << ", d2(qf)/d(p)d(p) = " << fwd_adj_p.nonzeros() << std::endl;

        // Adjoint over adjoint to get the second order sensitivities
        Function I_aoa = I_adj.factory("I_aoa", {"x0", "p", "adj_xf", "adj_qf", "adj:adj_p"},
                                                {"adj:x0", "adj:p"});
        arg = decltype(arg){{"x0", test.x0},
                            {"p", test.u0},
                            {"adj_xf", 0},
                            {"adj_qf", adj_qf},
                            {"adj_adj_p", 1}};
        res = I_aoa(arg);
        DM adj_adj_x0 = res.at("adj_x0");
        DM adj_adj_p = res.at("adj_p");
        std::cout << std::setw(50) << "Adjoint over adjoint sensitivities: " << "d2(qf)/d(x0)d(p) = " << adj_adj_x0.nonzeros() << ", d2(qf)/d(p)d(p) = " << adj_adj_p.nonzeros() << std::endl;
      }
    }
  }
  return 0;
}

File: test/matlab/sensitivity_analysis.m
%
%     This file is part of CasADi.
%
%     CasADi -- A symbolic framework for dynamic optimization.
%     Copyright (C) 2010-2014 Joel Andersson, Joris Gillis, Moritz Diehl,
%                             K.U. Leuven. All rights reserved.
%     Copyright (C) 2011-2014 Greg Horn
%
%     CasADi is free software; you can redistribute it and/or
%     modify it under the terms of the GNU Lesser General Public
%     License as published by the Free Software Foundation; either
%     version 3 of the License, or (at your option) any later version.
%
%     CasADi is distributed in the hope that it will be useful,
%     but WITHOUT ANY WARRANTY; without even the implied warranty of
%     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
%     Lesser General Public License for more details.
%
%     You should have received a copy of the GNU Lesser General Public
%     License along with CasADi; if not, write to the Free Software
%     Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
%
%

import casadi.*
disp 'Testing sensitivity analysis in CasADi'

% All ODE and DAE integrators to be tested
DAE_integrators = {'idas','collocation','oldcollocation'};
ODE_integrators = {'cvodes','rk', DAE_integrators{:}};

for ode=0:1
  if ode
    disp '******'
    disp 'Testing ODE example'
    Integrators = ODE_integrators;

    % Time 
    t = SX.sym('t');
      
    % Parameter
    u = SX.sym('u');

    % Differential states
    s = SX.sym('s'); v = SX.sym('v'); m = SX.sym('m');
    x = [s;v;m];

    % Constants
    alpha = 0.05; % friction
    beta = 0.1;   % fuel consumption rate
      
    % Differential equation
    ode = [v;
           (u-alpha*v*v)/m;
           -beta*u*u];
      
    % Quadrature
    quad = v^3 + ((3-sin(t)) - u)^2;

    % DAE
    dae = struct('t', t, 'x', x, 'p', u, 'ode', ode, 'quad', quad);

    % Time length
    tf = 0.5;

    % Initial position
    x0 = [0;0;1];

    % Parameter
    u0 = 0.4;

  else
    disp '******'
    disp 'Testing DAE example'    
    Integrators = DAE_integrators;

    % Differential state
    x = SX.sym('x');
    
    % Algebraic variable
    z = SX.sym('z');
    
    % Parameter
    u = SX.sym('u');
    
    % Differential equation
    ode = -x + 0.5*x*x + u + 0.5*z;
    
    % Algebraic constraint
    alg = z + exp(z) - 1.0 + x;
    
    % Quadrature
    quad = x*x + 3.0*u*u;

    % DAE
    dae = struct('x', x, 'z', z, 'p', u, 'ode', ode, 'alg', alg, 'quad', quad);
    
    % End time
    tf = 5;
    
    % Initial position
    x0 = 1;

    % Parameter
    u0 = 0.4;
  end

  % Integrator
  for integ=1:numel(Integrators)
    MyIntegrator = Integrators{integ};

    disp(sprintf('========'));
    disp(sprintf('Integrator: %s', MyIntegrator));
    disp(sprintf('========'));

    % Integrator options
    opts = struct;
    opts.tf = tf;
    if strcmp(MyIntegrator,'collocation') | strcmp(MyIntegrator,'oldcollocation')
      opts.implicit_solver = 'kinsol';
      opts.implicit_solver_options = struct('linear_solver', 'csparse');
      if strcmp(MyIntegrator,'oldcollocation')
        opts.expand_f=true;
      end
    end

    % Integrator
    I = Function.ivpsol('myintegrator', MyIntegrator, dae, opts);

    % Integrate to get results
    arg = struct;
    arg.x0 = x0;
    arg.p = u0;
    res = I(arg);
    xf = full(res.xf);
    qf = full(res.qf);
    fprintf('Unperturbed solution: xf=%d, qf=%d\n', xf, qf);

    % Perturb solution to get a finite difference approximation
    h = 0.001;
    arg.p = u0+h;
    res = I(arg);
    xf_pert = full(res.xf);
    qf_pert = full(res.qf);
    fprintf('Finite differences: d(xf)/d(p)=%d, d(qf)/d(p)=%d\n',...
        (xf_pert-xf)/h, (qf_pert-qf)/h);

    % Calculate once directional derivative, forward mode
    I_fwd = I.forward(1);
    arg = struct('der_x0',x0,'der_p',u0,'der_xf',xf,'der_qf',qf);
    arg.fwd0_x0 = 0;
    arg.fwd0_p = 1;
    res = I_fwd(arg);
    fwd_xf = full(res.fwd0_xf);
    fwd_qf = full(res.fwd0_qf);
    fprintf('Forward sensitivities: d(xf)/d(p)=%d, d(qf)/d(p)=%d\n', ...
        fwd_xf, fwd_qf);

    % Calculate one directional derivative, reverse mode
    I_adj = I.reverse(1);
    arg = struct('der_x0',x0,'der_p',u0,'der_xf',xf,'der_qf',qf);
    arg.adj0_xf = 0;
    arg.adj0_qf = 1;
    res = I_adj(arg);
    adj_x0 = full(res.adj0_x0);
    adj_p = full(res.adj0_p);
    fprintf('Adjoint sensitivities: d(qf)/d(x0)=%d, d(qf)/d(p)=%d\n', ...
        adj_x0, adj_p);

    % Perturb adjoint solution to get a finite difference approximation of
    % the second order sensitivities
    arg.der_p = u0+h;
    res = I_adj(arg);
    adj_x0_pert = full(res.adj0_x0);
    adj_p_pert = full(res.adj0_p);
    fprintf('FD of adjoint sensitivities: d2(qf)/d(x0)d(p)=%d, d2(qf)/d(p)d(p)=%d\n', ...
        (adj_x0_pert-adj_x0)/h, (adj_p_pert-adj_p)/h);

    % Forward over adjoint to get the second order sensitivities
    I_foa = I_adj.forward(1);
    arg = struct('der_der_x0',x0,'der_der_p',u0,'der_der_xf',xf,'der_der_qf',qf);
    arg.der_adj0_x0 = adj_x0;
    arg.der_adj0_p = adj_p;
    arg.der_adj0_xf = 0;
    arg.der_adj0_qf = 1;
    arg.fwd0_der_p = 1.0;
    res = I_foa(arg);
    fwd_adj_x0 = full(res.fwd0_adj0_x0);
    fwd_adj_p = full(res.fwd0_adj0_p);
    fprintf('Forward over adjoint sensitivities: d2(qf)/d(x0)d(p)=%d, d2(qf)/d(p)d(p)=%d\n', ...
        fwd_adj_x0, fwd_adj_p);

    % Adjoint over adjoint to get the second order sensitivities
    I_aoa = I_adj.reverse(1);
    arg = struct('der_der_x0',x0,'der_der_p',u0,'der_der_xf',xf,'der_der_qf',qf);
    arg.der_adj0_x0 = adj_x0;
    arg.der_adj0_p = adj_p;
    arg.der_adj0_xf = 0;
    arg.der_adj0_qf = 1;
    arg.adj0_adj0_p = 1;
    res = I_aoa(arg);
    adj_adj_x0 = full(res.adj0_der_x0);
    adj_adj_p = full(res.adj0_der_p);
    fprintf('Adjoint over adjoint sensitivities: d2(qf)/d(x0)d(p)=%d, d2(qf)/d(p)d(p)=%d\n', ...
        adj_adj_x0, adj_adj_p);
  end
end

